modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java

```diff
@@ -237,9 +266,41 @@ private boolean read(long pageId, ByteBuffer pageBuf, boolean checkCrc, boolean
         }
     }
 
+    /**
+     * Writes a page unlike {@link #write(long, ByteBuffer, boolean)}, checks the page offset in the file not logically (pageOffset <=
+     * {@link #pages()} * {@link #pageSize}) but physically (pageOffset <= {@link #size()}), which can affect performance when used in
+     * production code.
+     *
+     * @param pageId Page ID.
+     * @param pageBuf Page buffer to write from.
+     * @param calculateCrc If {@code false} crc calculation will be forcibly skipped.
+     * @throws IgniteInternalCheckedException If page writing failed (IO error occurred).
+     */
+    public void writeByPhysicalOffset(long pageId, ByteBuffer pageBuf, boolean calculateCrc) throws IgniteInternalCheckedException {
```

Comment:
This one seems very non-intuitive for me. We could allocate meta page before writing it, it shouldn't contradict anything. I don't really understand the necessity of 2 separate write methods in this case

Reply:
I think it's not right to reserve something with a **FilePageStore**, it should be responsible only for its functionality.

Reply:
That's correct.
I meant that a place for a meta page should be allocated before writing that meta page into a file. This way we wouldn't violate any internal assertions. And we _probably_ wouldn't need to manually set a page count to 1. Which is good.

Reply:
Fix it.